src/constraint_system/boolean.rs

```diff
@@ -50,6 +51,51 @@ where
 
         a
     }
+
+    /// A gate which outputs a variable whose value is 1 if
+    /// the input is 0 and whose value is 0 otherwise
+    pub fn is_zero_gate(&mut self, a: Variable) -> Variable {
+        //Get relevant field values
+        let a_value = self.variables.get(&a).unwrap();
+        let y_value = a_value.inverse().unwrap_or_else(E::Fr::one);
+        //This has value 1 if input value is zero, value 0 otherwise
+        let b_value = E::Fr::one() - *a_value * y_value;
+
+        let y = self.add_input(y_value);
+        let b = self.add_input(b_value);
+        let zero = self.zero_var();
+
+        // Enforce constraints. The constraint system being used here is
+        // a * y + b - 1 = 0
+        // a * b = 0
+        // where y is auxiliary and b is the boolean (a == 0).
+        let _a_times_b = self.arithmetic_gate(|gate| {
+            gate.witness(a, b, Some(zero)).mul(E::Fr::one())
+        });
+
+        let _constraint1 = self.arithmetic_gate(|gate| {
+            gate.witness(a, y, Some(zero))
+                .mul(E::Fr::one())
+                .fan_in_3(E::Fr::one(), b)
+                .constant(-E::Fr::one())
+        });
+
+        b
+    }
+
+    /// A gate which outputs a variable whose value is 1 if the
+    /// two input variables have equal values and whose value is 0 otherwise.
+    pub fn is_eq_gate(&mut self, a: Variable, b: Variable) -> Variable {
+        let difference = self.arithmetic_gate(|gate| {
+            gate.witness(a, b, None).add(E::Fr::one(), -E::Fr::one())
+        });
+        // TODO: These dummy constraints are a hack to avoid a 2^k issue,
+        // specifically the q_arith poly being constant.
+        // It should be safe to remove this dummy constraint once that issue
+        // is resolved.
+        self.add_dummy_constraints();
```

Comment:
This should be removed. 
Dummy constraints are only required for tests.
Anyway, the Composer by default adds some always (when you instanciate it) but I belive after @davidnevadoc 's PR, we don't have enough. 

This should be fixed, but not by adding the dummy constraints here.

Reply:
From @lopeetall 's comment above:

> Ok, mystery solved. This is because: `add_dummy_constraints` adds 2 arith gates to every circuit, and your circuit contains 6 more arith gates. This creates a circuit with 8 gates, all arith. This circuit is never padded out with zeros because it has a length of a power of 2 already. So the `q_arith` selector ends up as all 1s and its polynomial representation is degree 0 which triggers an error in KZG.
> 
> The original circuit had length 9. So it gets padded out to 16 with 0s and `q_arith` does not end up constant, so it works.
> 
> Indeed, if you make all the changes I suggested the total circuit length drops to 6. Then the test passes again!
> 
> I'll file an issue for the dummy constraint problem. It should be an easy fix.

So yes, this is a bad way to fix the problem.  Until the true problem is fixed this works, however.

Reply:
I should add here that the only problem with lopetall's comment is that the total circuit length drops to 6 within `is-zero-gate` but seems to then be 8 in `is-eq-gate`, hence the need to change the total number of gates there

Reply:
It should be easy to solve. And I would encourage you to do so. 

In order to solve it and make this PR clean in that regard, the solution would be to add the call to `add_dummy_constraints` 8 times when you instanciate a new `StandardComposer`.

This is done here:
https://github.com/ZK-Garage/plonk/blob/9e7c26539816bf33873cbeb51672da98f6a50e85/plonk-core/src/constraint_system/composer.rs#L221

Reply:
Resolved by #83 , the extra dummy constraints have been removed.